docs/why-and-how-of-inline-directives.md

```diff
@@ -0,0 +1,369 @@
+
+```purs
+foo = do
+  let
+    a = { bar: 42, baz: "something" }
+    b = 1 + f a
+  b
+```
+
+**To summarize, a data producer followed immediately by a data consumer indicates a place where an unneeded intermediate data structure exists and can be removed.**
+
+### Removing Unneeded Data Structures
+
+#### Inlining and Inline Directives
+
+An inliner replaces a function's call site (e.g. stuff on the left-hand-side of the `=`) with its implementation (e.g. stuff on the right-hand-side of the `=`). This replacement means the function's body is duplicated and will appear at least twice in the resulting code: once in its original definition and once in the usage site.
+
+Here's one example using the function `binaryPlus`:
+
+```purs
+binaryPlus :: Int -> Int -> Int
+binaryPlus a b = a + b
+
+usage = 9 + binaryPlus 1 2
+```
+
+There's a few ways we could inline the expression, `binaryPlus 1 2`. While `purs-backend-es` will performance its own analysis to determine whether and when to inline it, it doesn't know as much as the developer does.
+
+A developer can use **inline directives** to tell `purs-backend-es` whether to inline a function's body to its usage site and when (i.e. how many args need to be passed to the function before the inlining occurs):
+
+- never inline it at all (i.e. the directive is `never`)
+    - `9 + binaryPlus 1 2`
+- always inline it with no consideration for the number of arguments applied to it (i.e. the directive is `always`)
+    - `9 + (\a b -> a + b) 1 2`
+- inline it only after at least one argument has been applied to it (i.e. the directive is `arity=1`)
+    - `9 + (\  b -> 1 + b)   2`
+- inline it only after at least two arguments have been applied to it (i.e. the directive is `arity=2`)
+    - `9 + (        1 + 2)`
+
+**To summarize, inline directives provide developers with a stronger guarantee about whether and when a function is inlined than just relying upon the analysis done by `purs-backend-es`.**
+
+#### Inlining Duplicates Code
+
+The below example highlights the risk of inlining everything without thought: pointless code duplication.
+
+```purs
+hasExpensiveLetBinding a b = do
+  let
+    someValue = expensiveComputation
+  1 + someValue * b + a
+
+usage1 = hasExpensiveLetBinding 1 2
+usage2 = hasExpensiveLetBinding 3 4
+```
+
+If we inline `hasExpensiveLetBinding` immediately, the `expensiveComputation` will be computed three times rather than once. In other words, it would be the same as writing the following in source code:
+
+```purs
+hasExpensiveLetBinding a b = do
+  let
+    someValue = expensiveComputation
+  1 + someValue * b + a
+
+usage1 = (\a b -> do
+    let
+      someValue = expensiveComputation
+    1 + someValue * b + a
+  ) 1 2
+usage2 = (\a b -> do
+    let
+      someValue = expensiveComputation
+    1 + someValue * b + a
+  ) 3 4
+```
+
+**To summarize, inlining for the sake of inlining is a great way to unnecessarily increase your program's bundle size**.
+
+#### Rewrite Rules
```

Comment:
While the name of this section is not inaccurate, it's really confusing if you are familiar with GHC rewrite rules. "Rewrite rules" don't appear anywhere in the source code. So, yes, technically there are rules for how the optimizer rewrites things, but I would steer very clear of calling it that.

Reply:
Instead I would probably just call this evaluation. The optimizer knows how to evaluate PureScript code, and will evaluate things at build time.

Reply:
> I would steer very clear of calling it that.

Ok, I'll do that.

> will evaluate things at build time.

This isn't quite correct either as there are still some things it won't evaluate. For example:
```purs
mapWithIndex Tuple [ 1, 2, 3]
```
doesn't get evaluated to `[ Tuple 0 1, Tuple 1 2, Tuple 2 3 ]` because no such "rule" exists. However, calling it something else like "evaluation rules", which may be more accurate, is too similar to "rewrite rules".

Reply:
All I'm saying is that we shouldn't invent terminology that doesn't exist in the project's code base, and doesn't co-opt terminology used in another similar, high-profile project. Evaluation is what's happening, which is why the main function is called `eval`.

Reply:
How about this rendering?

> `purs-backend-es` _can_ evaluate PureScript code, but it will only evaluate _some_ expressions at build time. Typically, such expressions involve literal values (e.g. `1`, `{ foo: "string" }`). For example, primitive addition can be evaluated when its arguments are both literal `Int` values (e.g. `1 + 2` is evaluated to `3`). The below code demonstrates this: